core/store/src/trie/mem/arena/concurrent.rs

```diff
@@ -0,0 +1,252 @@
+/// normal arena, except that deallocation is not supported.
+///
+/// The only synchronization they need is the chunk counter. The purpose is so that after
+/// multiple threads allocate on their own arenas, the resulting memory can still be combined
+/// into a single arena while allowing the pointers (ArenaPos) to still be valid. This is what
+/// allows a memtrie to be loaded in parallel.
+pub struct ConcurrentArena {
+    /// Chunks allocated by
+    next_chunk_pos: Arc<AtomicUsize>,
+}
+
+impl ConcurrentArena {
+    pub fn new() -> Self {
+        Self { next_chunk_pos: Arc::new(AtomicUsize::new(0)) }
+    }
+
+    /// Returns an arena that can be used for one thread.
+    pub fn for_thread(&self) -> ConcurrentArenaForThread {
+        ConcurrentArenaForThread::new(self.next_chunk_pos.clone())
+    }
+
+    /// Converts the arena to a single-threaded arena. All returned values of `for_thread` must be
+    /// passed in.
+    ///
+    /// There is a caveat that may be fixed in the future if desired: the returned arena will have
+    /// some memory wasted. This is because the last chunk of each thread may not be full, and
+    /// the single-threaded arena is unable to make use of multiple partially filled chunks. The
+    /// maximum memory wasted is 4MB * number of threads; the average is 2MB * number of threads.
+    /// The wasted memory will be reclaimed when the memtrie shard is unloaded.
+    pub fn to_single_threaded(
+        self,
+        name: String,
+        threads: Vec<ConcurrentArenaForThread>,
+    ) -> STArena {
+        let mut chunks = Vec::new();
+        let mut active_allocs_bytes = 0;
+        let mut active_allocs_count = 0;
+        for thread in threads {
+            let memory = thread.memory;
+            let mut chunk_pos = vec![usize::MAX; memory.chunks.len()];
+            for (pos, index) in memory.chunk_pos_to_index.into_iter().enumerate() {
+                if index == usize::MAX {
+                    continue;
+                }
+                chunk_pos[index] = pos;
+            }
+            for (pos, chunk) in chunk_pos.into_iter().zip(memory.chunks.into_iter()) {
+                assert_ne!(pos, usize::MAX, "Not all arena threads have been passed in");
+                while chunks.len() <= pos {
+                    chunks.push(Vec::new());
+                }
+                chunks[pos] = chunk;
+            }
+            active_allocs_bytes += thread.allocator.active_allocs_bytes;
+            active_allocs_count += thread.allocator.active_allocs_count;
+        }
+        STArena::new_from_existing_chunks(name, chunks, active_allocs_bytes, active_allocs_count)
+    }
+}
+
+/// Arena to be used for a single thread.
+pub struct ConcurrentArenaForThread {
+    memory: ConcurrentArenaMemory,
+    allocator: ConcurrentArenaAllocator,
+}
+
+pub struct ConcurrentArenaMemory {
+    chunks: Vec<Vec<u8>>,
+    chunk_pos_to_index: Vec<usize>,
```

Comment:
Could you mention that this maps global chunk position to local chunk position, if local chunk belongs here, otherwise it is MAX? I was wondering why it is not HashMap until I looked at implementation.

Reply:
Renamed and commented. Thanks